prescient/data/providers/dat_data_provider.py

```diff
@@ -0,0 +1,296 @@
+        '''
+        self._populate_with_forecastable_data(options, start_time, num_time_periods,
+                                              time_period_length_minutes, model,
+                                              self._get_actuals_by_date)
+
+    def _populate_with_forecastable_data(self, options:Options,
+                                         start_time:datetime,
+                                         num_time_periods: int,
+                                         time_period_length_minutes: int,
+                                         model: EgretModel,
+                                         identify_dat: Callable[[date], EgretModel]
+                                        ) -> None:
+        # For now, require the time period to always be 60 minutes
+        assert(time_period_length_minutes == 60.0)
+        step_delta = timedelta(minutes=time_period_length_minutes)
+
+        # See if we have space to store all the requested data.
+        # If not, only supply what we have space for
+        if len(model.data['system']['time_keys']) < num_time_periods:
+            num_time_periods = len(model.data['system']['time_keys'])
+
+        # Collect a list of non-dispatchable generators
+        renewables = list(model.elements('generator', generator_type='renewable'))
+
+        start_hour = start_time.hour
+        start_day = start_time.date()
+
+        # Loop through each time step
+        for step_index in range(0, num_time_periods):
+            step_time = start_time + step_delta*step_index
+            day = step_time.date()
+
+            # 0-based hour, useable as index into forecast arrays
+            hour = step_time.hour
+
+            # For data starting at time 0, we collect tomorrow's data
+            # from today's dat file
+            if start_hour == 0 and day != start_day:
+                day = start_day
+                hour += 24
+
+            # If request is beyond the last day, just repeat the final day's values
+            if day > self._final_day:
+                day = self._final_day
+
+            dat = identify_dat(day)
+
+            # fill in renewables limits
+            for gen, gdata in renewables:
+                pmin = dat.data['elements']['generator'][gen]['p_min']['values'][hour]
+                pmax = dat.data['elements']['generator'][gen]['p_max']['values'][hour]
+                gdata['p_min']['values'][step_index] = pmin
+                gdata['p_max']['values'][step_index] = pmax
+
+            # Fill in load data
+            for bus, bdata in model.elements('load'):
+                load = dat.data['elements']['load'][bus]['p_load']['values'][hour]
+                bdata['p_load']['values'][step_index] = load
+
+            # Fill in reserve data
+            reserve_req = dat.data['system']['reserve_requirement']['values'][hour]
+            model.data['system']['reserve_requirement']['values'][step_index] = reserve_req
```

Comment:
Eventually we should make this logic somewhat generic, e.g., see https://github.com/grid-parity-exchange/Egret/blob/8a396c9191a5c8905ca93c3f9cf9b9f223a10d73/egret/model_library/transmission/tx_utils.py#L205

Reply:
Right now there are 5 places that have to know what is being predicted by a forecast.  I like the idea of reducing this down to a single spot.  Preferably I'd like to maintain the readability of `egret_plugin._copy_time_perid_data()`.  Also, of those 5 places I mentioned, 3 of them would require some non-trivial rework to get it to use a centralized definition of forecast data (creating a sced instance, storing forecasts, and storing forecast error).  For those reasons I'm deferring this to a new issue (issue #48)